### PR TITLE
Fix more alloc-dealloc mismatches and use-after-scopes

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fix another dynamically-sized allocation to use new/delete instead of the mismatched new[]/delete.

Move another local up a scope.

Use a simple new[]/delete[] instead of casting.

Discovered in https://github.com/dotnet/runtime/pull/54580

### CODE REVIEW

File: src/coreclr/vm/threadstatics.cpp [hunk 1]
```diff
@@ -136,7 +136,7 @@ void ThreadLocalBlock::EnsureModuleIndex(ModuleIndex index)
 
     // If this allocation fails, we will throw. If it succeeds,
     // then we are good to go
-    PTR_TLMTableEntry pNewModuleSlots = (PTR_TLMTableEntry) (void*) new BYTE[sizeof(TLMTableEntry) * aModuleIndices];
+    PTR_TLMTableEntry pNewModuleSlots = (PTR_TLMTableEntry) new TLMTableEntry[aModuleIndices];
```
Comment (jkotas) [inline src/coreclr/vm/threadstatics.cpp:139]:
```suggestion
    PTR_TLMTableEntry pNewModuleSlots = new TLMTableEntry[aModuleIndices];
```
This cast should be unnecessary now.

**Review by jkotas — APPROVED** [review]
Thanks!

File: src/coreclr/jit/lsrabuild.cpp [hunk 2]
```diff
@@ -3543,6 +3543,7 @@ int LinearScan::BuildReturn(GenTree* tree)
                 noway_assert(op1->IsMultiRegCall() || op1->IsMultiRegLclVar());
 
                 int                   srcCount;
+                ReturnTypeDesc        nonCallRetTypeDesc;
```
Comment (sandreenko) [inline src/coreclr/jit/lsrabuild.cpp:3546]: > Move another local up a scope.

what was the error here?

Comment (jkoritzinsky) [reply src/coreclr/jit/lsrabuild.cpp:3546]: The address of the old `retTypeDesc` was assigned to `pRetTypeDesc`, which was then used after the scope (ends on line 3561) on line 3562

**Review by sandreenko — APPROVED** [review]

**Pull request merged**